woocommerce/wc-admin
# Customers report: Name Advanced Filter

Fixes #1029.
Fixes #1084.

Adds the Name Advanced Filter to the _Customers_ report and a search box to the _Customers_ table.

The `<TableCard>` component was extended with two new props: `searchBy` and `searchParam`.

### Screenshots
![image](https://user-images.githubusercontent.com/3616980/50016994-16c12200-ffcb-11e8-9291-827742587a2d.png)

### Detailed test instructions:
There are several things to test here.

**Test 1**
- Go to the _Customers_ report without any filter applied `wc-admin#/analytics/customers`.
- Select _Advanced filters_ under the filter dropdown and _Add filter_ of type _Name_.
- Type any letter in the text box and verify suggestions appear.

**Test 2**
- Go to the _Customers_ report without any filter applied `wc-admin#/analytics/customers`.
- Search for a customer name in the table search box.
- Verify suggestions appear and, when selecting one, it's added to the filter.

**Test 3**
- Go to the _Customers_ report without any filter applied `wc-admin#/analytics/customers`.
- Select _Advanced filters_ under the filter dropdown and _Add filter_ of type _Name_.
- Write any name and click _Filter_.
- Search for a customer name in the table search box
- Verify suggestions appear but the current selected customer in the filters is not in the list. (In the screenshot above, notice _Aditya Jones_ doesn't appear as a suggestion).
- Verify that, when selecting another customer, it's added to the filter without deleting the previous one.

**Test 4**
- Test there are no regressions in other filters/reports.

### Future work
- The table search box will also have to search by email. But since that would have increased this PR complexity, I left it for now.
